archive/tar_stream.ts

```diff
@@ -0,0 +1,450 @@
+   * @default {''}
+   */
+  gname?: string;
+  /**
+   * The major number for character device.
+   * @default {''}
+   */
+  devmajor?: string;
+  /**
+   * The minor number for block device entry.
+   * @default {''}
+   */
+  devminor?: string;
+}
+
+const SLASH_CODE_POINT = "/".charCodeAt(0);
+
+/**
+ * ### Overview
+ * A TransformStream to create a tar archive.  Tar archives allow for storing
+ * multiple files in a single file (called an archive, or sometimes a tarball).
+ *   These archives typically have a single '.tar' extension.  This
+ * implementation follows the [FreeBSD 15.0](https://man.freebsd.org/cgi/man.cgi?query=tar&sektion=5&apropos=0&manpath=FreeBSD+15.0-CURRENT) spec.
+ *
+ * ### File Format & Limitations
+ * The ustar file format is used for creating the tar archive.  While this
+ * format is compatible with most tar readers, the format has several
+ * limitations, including:
+ * - Paths must be at most 256 characters.
+ * - Files must be at most 8 GiBs in size, or 64 GiBs if `sizeExtension` is set
+ * to true.
+ * - Sparse files are not supported.
+ *
+ * ### Usage
+ * TarStream may throw an error for several reasons. A few of those are:
+ * - The path is invalid.
+ * - The size provided does not match that of the iterable's length.
+ *
+ * ### Compression
+ * Tar archives are not compressed by default.  If you'd like to compress the
+ * archive, you may do so by piping it through a compression stream.
+ *
+ * @example Usage
+ * ```ts no-eval
+ * import { TarStream, type TarStreamInput } from "@std/archive/tar-stream";
+ *
+ * await ReadableStream.from<TarStreamInput>([
+ *   {
+ *     path: 'potato/'
+ *   },
+ *   {
+ *     path: 'deno.json',
+ *     size: (await Deno.stat('deno.json')).size,
+ *     readable: (await Deno.open('deno.json')).readable
+ *   },
+ *   {
+ *     path: '.vscode/settings.json',
+ *     size: (await Deno.stat('.vscode/settings.json')).size,
+ *     readable: (await Deno.open('.vscode/settings.json')).readable
+ *   }
+ * ])
+ *   .pipeThrough(new TarStream())
+ *   .pipeThrough(new CompressionStream('gzip'))
+ *   .pipeTo((await Deno.create('./out.tar.gz')).writable)
+ * ```
+ */
+
+export class TarStream implements TransformStream<TarStreamInput, Uint8Array> {
+  #readable: ReadableStream<Uint8Array>;
+  #writable: WritableStream<TarStreamInput>;
+  constructor() {
+    const { readable, writable } = new TransformStream<
+      TarStreamInput,
+      TarStreamInputInternal
+    >({
+      transform(chunk, controller) {
+        if (chunk.options) {
+          try {
+            assertValidTarStreamOptions(chunk.options);
+          } catch (e) {
+            return controller.error(e);
+          }
+        }
+
+        if (
+          "size" in chunk &&
+          (chunk.size < 0 || 8 ** 12 < chunk.size ||
+            chunk.size.toString() === "NaN")
+        ) {
+          return controller.error(new RangeError("Size cannot exceed 64 Gibs"));
```

Comment:
@kt3k, can you please go through and edit the error messages for this PR, as you see fit? As to keep things in line with the error guidelines.

Reply:
Updated the error messages to align them better to the style guide